go/pools/smartconnpool/pool.go

```diff
@@ -236,29 +244,54 @@ func (pool *ConnPool[C]) Open(connect Connector[C], refresh RefreshCheck) *ConnP
 
 // Close shuts down the pool. No connections will be returned from ConnPool.Get after calling this,
 // but calling ConnPool.Put is still allowed. This function will not return until all of the pool's
-// connections have been returned.
+// connections have been returned or the default PoolCloseTimeout has elapsed
 func (pool *ConnPool[C]) Close() {
-	if pool.close == nil {
+	ctx, cancel := context.WithTimeout(context.Background(), PoolCloseTimeout)
+	defer cancel()
+
+	if err := pool.CloseWithContext(ctx); err != nil {
+		log.Errorf("failed to close pool %q: %v", pool.Name, err)
+	}
+}
+
+// CloseWithContext behaves like Close but allows passing in a Context to time out the
+// pool closing operation
+func (pool *ConnPool[C]) CloseWithContext(ctx context.Context) error {
+	pool.capacityMu.Lock()
+	defer pool.capacityMu.Unlock()
+
+	if pool.close == nil || pool.capacity.Load() == 0 {
 		// already closed
-		return
+		return nil
 	}
 
-	pool.SetCapacity(0)
+	err := pool.setCapacity(ctx, 0)
 
 	close(pool.close)
 	pool.workers.Wait()
 	pool.close = nil
+	return err
 }
 
 func (pool *ConnPool[C]) reopen() {
+	pool.capacityMu.Lock()
+	defer pool.capacityMu.Unlock()
+
 	capacity := pool.capacity.Load()
 	if capacity == 0 {
 		return
 	}
 
-	pool.Close()
-	pool.open()
-	pool.SetCapacity(capacity)
+	ctx, cancel := context.WithTimeout(context.Background(), PoolCloseTimeout)
+	defer cancel()
+
+	if err := pool.setCapacity(ctx, 0); err != nil {
+		log.Errorf("failed to reopen pool %q: %v", pool.Name, err)
+	}
```

Comment:
I assume the reason to setCapacity to 0 first is to reset in case we did leak something previously? Maybe worth a comment in code to document what the PR description says.